### PR TITLE
feat(detector): add metrics for health checks

### PR DESCRIPTION
**Pull request by fcuny-rbx**
We expose metrics for the health checks from the detector itself.
There's three new metrics added:
- npd_detector_check_error: the check failed to run - it can be a
  problem with the check itself or other issues on the host that prevent
  it to run
- npd_detector_problem_counter: counts how many times a specific check
  failed. The label "check" with the name of the check is set.
- npd_detector_problem_gauge: set to 1 if a specific check is affecting
  the host or not. This is set if the check is failing, and unset once
  it stops. This can be used to report active problems, while the
  counter can be used for historical context.

Any other labels should be set by the agent scraping the metrics.

### CODE REVIEW

**Comment by fcuny-rbx** [comment]
I started `npd` with a disk limit of 10GB: `fcuny/detector-check-metrics`

Getting the metrics from the aggregator shows all the tests at 0:
```
vagrant@vagrant:~/go/src/github.com/Roblox/nomad-node-problem-detector$ curl -s localhost:8083/v1/metrics/|grep npd_
# HELP npd_detector_problem_counter Number of time a specific health checks failed
# TYPE npd_detector_problem_counter counter
npd_detector_problem_counter{check="CPUUnderPressure"} 0
npd_detector_problem_counter{check="DiskUsageHigh"} 0
npd_detector_problem_counter{check="MemoryUnderPressure"} 0
# HELP npd_detector_problem_gauge If a specific check is affecting the host or not
# TYPE npd_detector_problem_gauge gauge
npd_detector_problem_gauge{check="CPUUnderPressure"} 0
npd_detector_problem_gauge{check="DiskUsageHigh"} 0
npd_detector_problem_gauge{check="MemoryUnderPressure"} 0
```

Next, checking the current disk utilization, and we're indeed below the 10GB limit:
```
vagrant@vagrant:~/go/src/github.com/Roblox/nomad-node-problem-detector$ df -h /
Filesystem                    Size  Used Avail Use% Mounted on
/dev/mapper/vagrant--vg-root   62G  3.8G   55G   7% /
```

Now I create a file to go over the limit:
```
vagrant@vagrant:~/go/src/github.com/Roblox/nomad-node-problem-detector$ sudo fallocate -l 10G /test.img
vagrant@vagrant:~/go/src/github.com/Roblox/nomad-node-problem-detector$ df -h /
Filesystem                    Size  Used Avail Use% Mounted on
/dev/mapper/vagrant--vg-root   62G   14G   45G  24% /
```

And we can see that the counter increases and the gauge is set to 1:
```
vagrant@vagrant:~/go/src/github.com/Roblox/nomad-node-problem-detector$ curl -s localhost:8083/v1/metrics/|grep npd_
# HELP npd_detector_problem_counter Number of time a specific health checks failed
# TYPE npd_detector_problem_counter counter
npd_detector_problem_counter{check="CPUUnderPressure"} 0
npd_detector_problem_counter{check="DiskUsageHigh"} 1
npd_detector_problem_counter{check="MemoryUnderPressure"} 0
# HELP npd_detector_problem_gauge If a specific check is affecting the host or not
# TYPE npd_detector_problem_gauge gauge
npd_detector_problem_gauge{check="CPUUnderPressure"} 0
npd_detector_problem_gauge{check="DiskUsageHigh"} 1
npd_detector_problem_gauge{check="MemoryUnderPressure"} 0
vagrant@vagrant:~/go/src/github.com/Roblox/nomad-node-problem-detector$ curl -s localhost:8083/v1/metrics/|grep npd_
# HELP npd_detector_problem_counter Number of time a specific health checks failed
# TYPE npd_detector_problem_counter counter
npd_detector_problem_counter{check="CPUUnderPressure"} 0
npd_detector_problem_counter{check="DiskUsageHigh"} 3
npd_detector_problem_counter{check="MemoryUnderPressure"} 0
# HELP npd_detector_problem_gauge If a specific check is affecting the host or not
# TYPE npd_detector_problem_gauge gauge
npd_detector_problem_gauge{check="CPUUnderPressure"} 0
npd_detector_problem_gauge{check="DiskUsageHigh"} 1
npd_detector_problem_gauge{check="MemoryUnderPressure"} 0
```

Deleting the file to bring it back under the 10GB limit:
```
vagrant@vagrant:~/go/src/github.com/Roblox/nomad-node-problem-detector$ sudo rm /test.img
vagrant@vagrant:~/go/src/github.com/Roblox/nomad-node-problem-detector$ curl -s localhost:8083/v1/metrics/|grep npd_
# HELP npd_detector_problem_counter Number of time a specific health checks failed
# TYPE npd_detector_problem_counter counter
npd_detector_problem_counter{check="CPUUnderPressure"} 0
npd_detector_problem_counter{check="DiskUsageHigh"} 4
npd_detector_problem_counter{check="MemoryUnderPressure"} 0
# HELP npd_detector_problem_gauge If a specific check is affecting the host or not
# TYPE npd_detector_problem_gauge gauge
npd_detector_problem_gauge{check="CPUUnderPressure"} 0
npd_detector_problem_gauge{check="DiskUsageHigh"} 0
npd_detector_problem_gauge{check="MemoryUnderPressure"} 0
```

We can see the counter reports the number of times the check failed, and the gauge is back at 0 since this is not affecting the host anymore.

**Review by vineesha92 — APPROVED** [review]

**Comment by danmaasrbx** [comment]
Suggestions on metric names (not urgent, just thinking about Prometheus conventions):

- "npd_detector_check_error" -> "npd_detector_check_error_count"
- "npd_detector_problem_counter" -> "npd_detector_problem_count"
  -  "_count" is more common than "_counter" for a metric like this
- "npd_detector_problem_gauge" -> "npd_detector_problem"
  - "gauge" is assumed by default when there is no "_count" suffix

**Comment by fcuny-rbx** [comment]
@danmaasrbx thanks for the feedback, I've renamed these metrics (do you have a recommendation regarding what I can read for best practices when it comes to naming ? Or is it mostly from experience that you've come to figure out what makes sense to name them ?)

File: detector/detector.go [hunk 1]
```diff
@@ -40,6 +40,13 @@ import (
 	"github.com/urfave/cli/v2"
 )
 
+const (
```
Comment (fcuny-rbx) [inline detector/detector.go:43]: this should be dropped, it's not used - I'll clean this up.

**Comment by fcuny-rbx** [comment]
@shishir-a412ed I forgot this one was still open, is there anything else you want me to address in this change ?

**Comment by shishir-a412ed** [comment]
@fcuny-rbx I haven't checked this in a while, will review.
Also, removing the `work-in-progress` label.

This PR might also need a rebase!

**Review by clobrox — APPROVED** [review]
Nice!

File: detector/detector.go (hunk 1 above)
Comment (shishir-a412ed) [reply detector/detector.go:43]: Are we using these constants or should we remove them?

Comment (fcuny-rbx) [reply detector/detector.go]: I dropped them, they were not used

**Comment by fcuny-rbx** [comment]
@shishir-a412ed I removed the unused constants and rebased on top of main - let me know if there's anything else I'm missing

File: detector/detector.go [hunk 2]
```diff
@@ -399,9 +419,32 @@ func metricsHandler(registry *prometheus.Registry) http.Handler {
 }
 
 func registerMetrics() *prometheus.Registry {
+	counterOpts := prometheus.CounterOpts{}
+	gaugeOpts := prometheus.GaugeOpts{}
+
+	if nomadDc, ok := os.LookupEnv("NOMAD_DC"); ok {
```
Comment (shishir-a412ed) [inline detector/detector.go:425]: Shall we do a check to cover the empty case?

```
if nomadDC, ok := os.LookupEnv("NOMAD_DC"); ok && nomadDC != "" {
   ...
}
```

While golang prefers camelCase for variable names, for acronyms like URL or HTTP, it's all caps e.g. ServeHTTP.
Shall we rename it to `nomadDC`? (Just a suggestion, feel free to ignore 🙂).

[`Reference blog post`](https://blog.devgenius.io/write-better-go-code-by-using-consistent-case-for-initials-in-variables-b84e862fd0ba)

**Review by shishir-a412ed — APPROVED** [review]
small nitpicks, otherwise LGTM.

Thanks @fcuny-rbx!

File: detector/detector.go (hunk 2 above)
Comment (fcuny-rbx) [reply detector/detector.go:425]: Thanks, I think this is more consistent, I've renamed the variable.

**Pull request merged**